src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs

```diff
@@ -109,14 +109,15 @@ public static partial class JsonSerializer
             IDictionary? extensionData = (IDictionary?)jsonPropertyInfo.GetValueAsObject(obj);
             if (extensionData == null)
             {
+#if DEBUG
```

Comment:
> so I wasn't sure if it was better to put this discussion in an issue or PR, but chose PR to better discuss the code changes

A PR is a fine choice to discuss code changes, particularly to fix an issue.

> should I expect that #if DEBUG would change the behavior of tests running in Debug configuration?

Simply wrapping the `Debug.Assert` calls as this PR does has no effect. They only run in debug mode to begin with.

This #if directive allows us to execute code that will only run in `debug` builds of System.Text.Json. The reason I had it in https://gist.github.com/layomia/d77aeea5d3250785de5a7aa4dee95c46#file-create_data_extension_property-cs-L8-L20 was to make the following assignments which we can use to perform assertions in debug mode.

```C#
Type underlyingIDictionaryType = jsonPropertyInfo.DeclaredPropertyType.GetCompatibleGenericInterface(typeof(IDictionary<,>))!;
Type[] genericArgs = underlyingIDictionaryType.GetGenericArguments();
```

We don't want to do this in release mode as these calls have a non-trivial perf cost.

`underlyingIDictionaryType` is the type of `IDictionary<,>` the extension data implements. This is the type for which we should perform the checks that the first generic parameter is `string`, and the second generic parameter is `object` or `JsonElement`. Performing these checks on the declared type is what leads to the `IndexOutOfRangeException` (because the assertions may not hold).

> Is there a reason to keep the Debug.Assert statements here at all?

The reason we want the assertions is to ensure that no code changes [upstream](https://github.com/dotnet/runtime/blob/0d3ed81b48d30c10ed213f087849359d029a0d3c/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L222) that violates our expectations of the type we think we are processing in this method.

> If conceptually System.Text.Json supports dictionaries that don't match the conditions tested by the Debug.Assert statements, what is their purpose?

This area of code is only concerned with initializing extension data dictionaries. We support dictionaries with string keys and values of any type, but extension properties must be of types assignable to `IDictionary<string, object/JsonElement>`.

Reply:
https://gist.github.com/layomia/d77aeea5d3250785de5a7aa4dee95c46#file-create_data_extension_property-cs-L10-L11

🤦‍♂  I completely missed that in the gist.  This whole thing makes 100% more sense.  Will fix.

Reply:
fixed in e398543

Now, I'm finding that dictionaries that implement `IDictionary<string,object>` but not `IDictionary` run into an `InvalidCastException` at 

https://github.com/dotnet/runtime/blob/e3985439e591cc1a3a0b8c51ccc13cdd3168cf20/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs#L130

The same cast is used beforehand at:
https://github.com/dotnet/runtime/blob/e3985439e591cc1a3a0b8c51ccc13cdd3168cf20/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs#L110-L111

Based on your comment above:
> We support dictionaries with string keys and values of any type, but extension properties must be of types assignable to `IDictionary<string, object/JsonElement>`.

Must these `ExtensionData` properties also implement `IDictionary` as a conceptual constraint, or is that something that should change

Reply:
Yes, we should remove the casts; particularly because they are to the wrong type. The requirement is `IDictionary<string,object/JsonElement>`. The `IDictionary` cast must have been an oversight.

We've already verified the type, and are only creating the object here; so we don't need to cast it.

Reply:
fixed in d19fff9